### PR TITLE
Issue 239 Fix

### PR DESCRIPTION
**Pull request by sageserpent-open**
Addresses this issue: [Possible bug in implementation of MapDecorator.mergeByKeyWith](https://github.com/scala/scala-collection-contrib/issues/239).

Test added to reproduce the bug in: ~~bcc2b6efedcb97ae329b792940b52e44ef1db06e~~4fe84ca7370f05059817fd333f72719d33fb4652.

Fix to make test pass in: ~~5088ed6ac8a23e3a9987491bf05f334d05cdfc74~~8c8d67bd323f7a9bf2b786b5b55c3349fa52f7f8.

### CODE REVIEW

**Comment by sageserpent-open** [comment]
The second commit is a bit messy - let me tidy up the history a little for the sake of clarity and force-push again...

**Comment by sageserpent-open** [comment]
Refreshed the PR and edited the original description accordingly.

**Comment by sageserpent-open** [comment]
@SethTisue , @Philippus This PR is awaiting approval from a maintainer. Could somebody take a look if they have a moment, please?

File: src/test/scala/scala/collection/decorators/MapDecoratorTest.scala
Comment (som-snytt) [inline src/test/scala/scala/collection/decorators/MapDecoratorTest.scala:116]: This compares keySet and keys. Does it matter? Without a resolution to https://github.com/scala/scala/pull/10766 it's fruitless to speculate.

**Review by som-snytt — APPROVED** [review]
Nice catch. Does this mean no one uses this code? Worth noting that the original is surprisingly hard to read, in part because `Option#fold` is obfuscating.

**Comment by som-snytt** [comment]
I see the PR is unsquashed, but I don't know what local policy is for this repo.

The test is interesting in that it demonstrates the old behavior. Is such a test meaningful after the fix? That is, it becomes a test where one asks sometime in the future, Why did they ever think this test could fail? It tests for apples instead of oranges.

**Comment by sageserpent-open** [comment]
> Nice catch. Does this mean no one uses this code? Worth noting that the original is surprisingly hard to read, in part because `Option#fold` is obfuscating.

Well, I certainly use the code, although in the meantime I’ve copied the fixed code locally.😊

**Comment by sageserpent-open** [comment]
> I see the PR is unsquashed, but I don't know what local policy is for this repo.

I was aiming to show the test failing first, then passing with the fix - I like to split my work into testing, implementing, refactoring, tidying up to avoid one big PR - although this is just a minnow anyway. It’s no problem from my side if you want to squash the merge if you’re happy with the PR.
> 
> The test is interesting in that it demonstrates the old behavior. Is such a test meaningful after the fix? That is, it becomes a test where one asks sometime in the future, Why did they ever think this test could fail? It tests for apples instead of oranges.

I was thinking of the test as being a bug reproduction test, but wrote it as two scenarios to show the simpler situation first where values don’t collide, followed by the one where they do.

Depending on how you philosophise about testing, you might say that the second case is a case of “let’s show that the SUT does not make my breakfast / make nuisance phone calls / launch nuclear missiles”, or instead you could say that preservation of the values is part of its wider contract.

I take the latter view, but if it’s a dealbreaker, how about we split the test? FWIW, I saw this repository as more of a fast-and-loose affair, so I just got something in to support the fix.

File: src/test/scala/scala/collection/decorators/MapDecoratorTest.scala
Comment (sageserpent-open) [reply src/test/scala/scala/collection/decorators/MapDecoratorTest.scala:116]: Oops - yes, let me fix this…

Comment (sageserpent-open) [reply src/test/scala/scala/collection/decorators/MapDecoratorTest.scala]: **DONE**

**Comment by sageserpent-open** [comment]
@som-snytt / @SethTisue Another bump for a workflow approval, please. Next time I won't fork the repository to submit a PR, lesson learned...

**Review by Philippus — APPROVED** [review]

**Pull request merged**
